radix-engine/src/blueprints/pool/single_resource_pool/blueprint.rs

```diff
@@ -0,0 +1,408 @@
+
+        api.globalize_with_address(
+            btreemap!(
+                ObjectModuleId::Main => object_id,
+                ObjectModuleId::AccessRules => access_rules.0,
+                ObjectModuleId::Metadata => metadata.0,
+                ObjectModuleId::Royalty => royalty.0,
+            ),
+            address,
+        )?;
+
+        Ok(ComponentAddress::new_or_panic(address.as_node_id().0))
+    }
+
+    pub fn contribute<Y>(
+        bucket: Bucket,
+        api: &mut Y,
+    ) -> Result<SingleResourcePoolContributeOutput, RuntimeError>
+    where
+        Y: ClientApi<RuntimeError>,
+    {
+        // No check that the bucket is of the same resource as the vault. This check will be handled
+        // by the vault itself on deposit.
+
+        let (substate, handle) = Self::lock_and_read(api, LockFlags::read_only())?;
+        let mut pool_unit_resource_manager = substate.pool_unit_resource_manager();
+        let mut vault = substate.vault();
+
+        if bucket.is_empty(api)? {
+            return Err(SingleResourcePoolError::ContributionOfEmptyBucketError.into());
+        }
+
+        /*
+        There are four states that the pool could be in at this point of time depending on the total
+        supply of the pool units and the the total amount of reserves that the pool unit has. We can
+        examine each of those states.
+
+        Let PU denote the total supply of pool units where 0 means that none exists and 1 means that
+        some amount exists. Let R denote the total amount of reserves that the pool has where 0 here
+        means that no reserves exist in the pool and 1 means that some reserves exist in the pool.
+
+        PU  R
+        0   0 => This is a new pool - no pool units and no pool reserves.
+        0   1 => This is a pool which has been used but has dried out and all of the pool units have
+                 been burned. The first contribution to this pool gets whatever dust is left behind.
+        1   0 => This is an illegal state! Some amount of people own some % of zero which is invalid
+        1   1 => The pool is in normal operations.
+
+        Thus depending on the supply of these resources the pool behaves differently.
+         */
+
+        let reserves = vault.amount(api)?;
+        let pool_unit_total_supply = pool_unit_resource_manager.total_supply(api)?;
+        let amount_of_contributed_resources = bucket.amount(api)?;
+
+        let pool_units_to_mint = match (
+            pool_unit_total_supply > Decimal::ZERO,
+            reserves > Decimal::ZERO,
+        ) {
+            (false, false) => Ok(amount_of_contributed_resources),
+            (false, true) => Ok(amount_of_contributed_resources + reserves),
+            (true, false) => Err(SingleResourcePoolError::NonZeroPoolUnitSupplyButZeroReserves),
+            (true, true) => Ok(amount_of_contributed_resources / reserves * pool_unit_total_supply),
```

Comment:
Do we lose precision here since we are dividing first?

Reply:
Yup, updated.